pkg/server/api/agent/v1/service.go

```diff
@@ -243,7 +257,128 @@ func (s *Service) BanAgent(ctx context.Context, req *agent.BanAgentRequest) (*em
+	}
+
+	// validate
+	params := req.GetParams()
+	if err := validateAttestAgentParams(params); err != nil {
+		log.WithError(err).Error("Invalid argument")
+		return status.Errorf(codes.InvalidArgument, "invalid argument: %v", err)
+	}
+
+	attestorType := params.Data.Type
+	log = log.WithField(telemetry.NodeAttestorType, attestorType)
+
+	// attest
+	var attestResp *nodeattestor.AttestResponse
+	if attestorType == "join_token" {
+		attestResp, err = s.attestJoinToken(ctx, params.Data.Payload)
+		if err != nil {
+			return err
+		}
+	} else {
+		attestResp, err = s.attestChallengeResponse(ctx, stream, params, attestorType)
+		if err != nil {
+			return err
+		}
+	}
+
+	agentID, err := spiffeid.FromString(attestResp.AgentId)
+	if err != nil {
+		log.WithError(err).Error("Invalid agent id")
+		return status.Error(codes.Internal, "invalid agent id")
+	}
+	log = log.WithField(telemetry.AgentID, agentID.String())
+
+	// fetch the agent/node to check if it was already attested or banned
+	attestedNode, err := s.ds.FetchAttestedNode(ctx, &datastore.FetchAttestedNodeRequest{
+		SpiffeId: agentID.String(),
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to fetch agent")
+		return status.Error(codes.Internal, "failed to fetch agent")
+	}
+
+	if isBanned(attestedNode.Node) {
+		log.WithError(err).Error("Failed to attest; agent is banned")
+		return status.Error(codes.PermissionDenied, "failed to attest; agent is banned")
+	}
+
+	// parse and sign CSR
+	svid, err := s.signSvid(ctx, &agentID, params.Params.Csr, log)
+	if err != nil {
+		return err
+	}
+
+	// augment selectors with resolver
+	augmentedSels, err := s.augmentSelectors(ctx, agentID.String(), attestResp.Selectors, params.Data.Type)
+	if err != nil {
+		log.WithError(err).Error("Failed to augment selectors")
+		return status.Error(codes.Internal, "failed to augment selectors")
+	}
+	// store augmented selectors
+	_, err = s.ds.SetNodeSelectors(ctx, &datastore.SetNodeSelectorsRequest{
+		Selectors: &datastore.NodeSelectors{
+			SpiffeId:  agentID.String(),
+			Selectors: augmentedSels,
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to update selectors")
+		return status.Error(codes.Internal, "failed to update selectors")
+	}
+
+	// create or update attested entry
+	if attestedNode.Node == nil {
+		req := &datastore.CreateAttestedNodeRequest{
+			Node: &common.AttestedNode{
+				AttestationDataType: params.Data.Type,
+				SpiffeId:            agentID.String(),
+				CertNotAfter:        svid[0].NotAfter.Unix(),
+				CertSerialNumber:    svid[0].SerialNumber.String(),
+			}}
+		if _, err := s.ds.CreateAttestedNode(ctx, req); err != nil {
+			log.WithError(err).Errorf("Failed to create attestation entry: %q", agentID)
+			return status.Error(codes.Internal, "failed to create attestation entry")
+		}
+	} else {
+		req := &datastore.UpdateAttestedNodeRequest{
+			SpiffeId:         agentID.String(),
```

Comment:
This will clean the fields `NewCertSerialNumber` and `NewCertNotAfter`. It's my understanding that is necessary, given that we are attesting once more and those fields used by the rotation are meaningless at this point.